File: src/main/java/hiiragi283/material/api/shape/HiiragiShapes.java
package hiiragi283.material.api.shape;

import hiiragi283.material.api.registry.HiiragiRegistryEntries;

public class HiiragiShapes extends HiiragiRegistryEntries {

    public static final HiiragiShape BALL = new HiiragiShape("ball", 30);
    public static final HiiragiShape BLOCK = new HiiragiShape("block", 144 * 9);
    public static final HiiragiShape BOTTLE = new HiiragiShape("bottle", 144);
    public static final HiiragiShape CASING = new HiiragiShape("casing", (144 * 4) + (144 * 6));
    public static final HiiragiShape CLUMP = new HiiragiShape("clump", 144);
    public static final HiiragiShape CLUSTER = new HiiragiShape("cluster", 144 * 2);
    public static final HiiragiShape COIN = new HiiragiShape("coin", 144 / 3);
    public static final HiiragiShape CRUSHED = new HiiragiShape("crushed", 144);
    public static final HiiragiShape CRYSTAL = new HiiragiShape("crystal", 144);
    public static final HiiragiShape DUST = new HiiragiShape("dust", 144);
    public static final HiiragiShape DUST_DIRTY = new HiiragiShape("dust_dirty", 144);
    public static final HiiragiShape DUST_TINY = new HiiragiShape("dust_tiny", 144 / 9);
    public static final HiiragiShape FENCE = new HiiragiShape("fence", 144 * 5 / 3);
    public static final HiiragiShape FRAME = new HiiragiShape("frame", 144 * 4);
    public static final HiiragiShape GEAR = new HiiragiShape("gear", 144 * 4);
    public static final HiiragiShape GEM = new HiiragiShape("gem", 144);
    public static final HiiragiShape INGOT = new HiiragiShape("ingot", 144);
    public static final HiiragiShape LOG = new HiiragiShape("log", 144 * 4);
    public static final HiiragiShape NUGGET = new HiiragiShape("nugget", 144 / 9);
    public static final HiiragiShape ORE = new HiiragiShape("ore", 144 * 2);
    public static final HiiragiShape ORE_DENSE = new HiiragiShape("ore_dense", 144 * 2);
    public static final HiiragiShape ORE_POOR = new HiiragiShape("ore_poor", 144 / 3);
    public static final HiiragiShape PLANK = new HiiragiShape("plank", 144);
    public static final HiiragiShape PLATE = new HiiragiShape("plate", 144);
    public static final HiiragiShape PLATE_DENSE = new HiiragiShape("plate_dense", 144 * 9);
    public static final HiiragiShape PURIFIED = new HiiragiShape("crushed_purified", 144);
    public static final HiiragiShape SAND = new HiiragiShape("sand", 144);
    public static final HiiragiShape SCAFFOLDING = new HiiragiShape("scaffolding", (144 * 4 + 72) / 6);
    public static final HiiragiShape SHARD = new HiiragiShape("shard", 144);
    public static final HiiragiShape SHEETMETAL = new HiiragiShape("block_sheetmetal", 144);
    public static final HiiragiShape SLAB = new HiiragiShape("slab", 144 * 9 / 2);
    public static final HiiragiShape SLAB_SHEETMETAL = new HiiragiShape("slab_sheetmetal", 144 / 2);
    public static final HiiragiShape STICK = new HiiragiShape("stick", 144 / 2);
    public static final HiiragiShape STONE = new HiiragiShape("stone", 144);
    public static final HiiragiShape WIRE = new HiiragiShape("wire", 144 / 2);

    //    State    //

    public static final HiiragiShape SOLID = new HiiragiShape("solid", 0);

    public static final HiiragiShape LIQUID = new HiiragiShape("liquid", 0);

    public static final HiiragiShape GAS = new HiiragiShape("gas", 0);

    //    Type    //

    public static final HiiragiShape METAL = new HiiragiShape("metal", 0);

}

File: src/main/kotlin/hiiragi283/material/api/shape/HiiragiShapes.kt
package hiiragi283.material.api.shape

import hiiragi283.material.api.material.CrystalType
import hiiragi283.material.api.material.MaterialRegistry
import hiiragi283.material.util.CraftingBuilder
import hiiragi283.material.util.RagiIngredient
import hiiragi283.material.util.append
import hiiragi283.material.util.toLocation
import net.minecraft.client.renderer.block.model.ModelResourceLocation
import net.minecraftforge.client.model.ModelLoader

object HiiragiShapes {

    @JvmField
    val COMMON: (Double) -> HiiragiShape = { shapeOf("common", it) }

    @JvmField
    val BALL = shapeOf("ball", 0.2)

    @JvmField
    val BLOCK = shapeOf(
        "block",
        9.0,
        model = {
            val common = ModelResourceLocation(it.getLocation()!!.append("_material"), "inventory")
            val gem = ModelResourceLocation(it.getLocation()!!.append("_gem"), "inventory")
            val metal = ModelResourceLocation(it.getLocation()!!.append("_metal"), "inventory")

            ModelLoader.registerItemVariants(it.getItem(), common, gem, metal)

            ModelLoader.setCustomMeshDefinition(it.getItem()) { stack ->
                val material = MaterialRegistry.getMaterial(stack.metadata)
                if (material.isMetal()) metal
                else if (material.isGem()) gem
                else common
            }
        },
        recipe = { entry, material ->
            if (material.isSolid()) {
                if (material.isGem()) {
                    CraftingBuilder(entry.getItemStack(material))
                        .setPattern("AAA", "AAA", "AAA")
                        .setIngredient('A', "gem${material.getOreDictName()}")
                        .buildShaped()
                } else {
                    CraftingBuilder(entry.getItemStack(material))
                        .setPattern("AAA", "AAA", "AAA")
                        .setIngredient('A', "ingot${material.getOreDictName()}")
                        .buildShaped()
                }
            }
        }
    )

    @JvmField
    val BOTTLE = shapeOf("bottle", 1.0)

    @JvmField
    val CLUMP = shapeOf("clump", 1.0)

    @JvmField
    val CLUSTER = shapeOf("cluster", 2.0)

    @JvmField
    val COIN = shapeOf("coin", 0.3)

    @JvmField
    val CRUSHED = shapeOf("crushed", 1.0)

    @JvmField
    val CRYSTAL = shapeOf("crystal", 1.0)

    @JvmField
    val DUST = shapeOf("dust", 1.0, recipe = { entry, material ->
        CraftingBuilder(entry.getItemStack(material))
            .setPattern("AAA", "AAA", "AAA")
            .setIngredient('A', "dustTiny${material.getOreDictName()}")
            .buildShaped()
    })

    @JvmField
    val DUST_DIRTY = shapeOf("dust_dirty", 1.0)

    @JvmField
    val DUST_TINY = shapeOf("dust_tiny", 0.1, recipe = { entry, material ->
        CraftingBuilder(entry.getItemStack(material, 9))
            .addIngredient(RagiIngredient("dust${material.getOreDictName()}"))
            .buildShapeless()
    })

    @JvmField
    val GEAR = shapeOf(
        "gear", 4.0,
        /*recipe = { item, material ->
               CraftingBuilder(item.getItemStack(material))
                   .setPattern(" A ", "A A", " A ")
                   .setIngredient('A', "ingot${material.getOreDictName()}")
                   .buildShaped()
           }*/
    )

    @JvmField
    val GEM = shapeOf(
        "gem",
        1.0,
        model = { entry ->
            ModelLoader.registerItemVariants(entry.getItem(), *CrystalType.values()
                .filter { it.texture.isNotEmpty() }
                .map { entry.getLocation()!!.append("_" + it.texture) }
                .toTypedArray())

            ModelLoader.setCustomMeshDefinition(entry.getItem()) { stack ->
                val material = MaterialRegistry.getMaterial(stack.metadata)
                val type = if (material.isGem()) material.crystalType else CrystalType.CUBIC
                type.getLocation(entry.getItem())
            }
        },
        recipe = { item, material ->
            CraftingBuilder(item.getItemStack(material, 9))
                .addIngredient(RagiIngredient("block${material.getOreDictName()}"))
                .buildShapeless()
        }
    )

    @JvmField
    val INGOT = shapeOf("ingot", 1.0, recipe = { entry, material ->
        //nugget -> ingot
        CraftingBuilder(entry.getItemStack(material))
            .setPattern("AAA", "AAA", "AAA")
            .setIngredient('A', "nugget${material.getOreDictName()}")
            .buildShaped()
        //block -> ingot
        val ingot9 = entry.getItemStack(material, 9)
        CraftingBuilder(ingot9.toLocation("_").append("_alt"), ingot9)
            .addIngredient(RagiIngredient("block${material.getOreDictName()}"))
            .buildShapeless()
    })

    @JvmField
    val LOG = shapeOf("log", 4.0)

    @JvmField
    val NUGGET = shapeOf("nugget", 0.1, recipe = { entry, material ->
        CraftingBuilder(entry.getItemStack(material, 9))
            .addIngredient(RagiIngredient("ingot${material.getOreDictName()}"))
            .buildShapeless()
    })

    @JvmField
    val ORE = shapeOf("ore", -1.0)

    @JvmField
    val ORE_POOR = shapeOf("ore_poor", -1.0)

    @JvmField
    val PLANK = shapeOf("plank", 1.0)

    @JvmField
    val PLATE = shapeOf(
        "plate", 1.0,
        /*recipe = { item, material ->
               if (material.isMetal()) {
                   CraftingBuilder(item.getItemStack(material))
                       .addIngredient(RagiIngredient("ingot${material.getOreDictName()}"))
                       .addIngredient(RagiIngredient(ItemStack(RMItems.FORGE_HAMMER, 1, OreDictionary.WILDCARD_VALUE)))
                       .buildShapeless()
               }
           }*/
    )

    @JvmField
    val PLATE_DENSE = shapeOf("plate_dense", 9.0)

    @JvmField
    val PURIFIED = shapeOf("crushed_purified", 1.0)

    @JvmField
    val SAND = shapeOf("sand", 1.0)

    @JvmField
    val SHARD = shapeOf("shard", 1.0)

    @JvmField
    val STICK = shapeOf(
        "stick", 0.5,
        /*recipe = { item, material ->
               if (material.isMetal()) {
                   CraftingBuilder(item.getItemStack(material, 4))
                       .setPattern("AB", "A ")
                       .setIngredient('A', "ingot${material.getOreDictName()}")
                       .setIngredient('B', ItemStack(RMItems.FORGE_HAMMER, 1, OreDictionary.WILDCARD_VALUE))
                       .buildShaped()
               }
           }*/
    )

    @JvmField
    val STONE = shapeOf("stone", 1.0)

    fun register(registry: MutableList<HiiragiShape>) {
        this::class.java.declaredFields
            .map { it.also { it.isAccessible = true } }
            .map { it.get(this) }
            .filterIsInstance<HiiragiShape>()
            .forEach { registry.add(it) }
    }

}